From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mumbel <22204578+mumbel@users.noreply.github.com>
Date: Sun, 21 Jan 2024 13:46:42 -0600
Subject: [PATCH] 6138: Xtensa: ELF Relocation

Fixed switch statement missing break
Implemented a few of the more important cases
---
 .../Xtensa_ElfRelocationConstants.java        |  20 +-
 .../Xtensa_ElfRelocationHandler.java          | 433 ++++++++++++------
 2 files changed, 300 insertions(+), 153 deletions(-)

diff --git a/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationConstants.java b/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationConstants.java
index e5b959849..35cb9908e 100644
--- a/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationConstants.java
+++ b/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationConstants.java
@@ -4,9 +4,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *      http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -18,7 +18,7 @@ package ghidra.app.util.bin.format.elf.relocation;
 public class Xtensa_ElfRelocationConstants {
 	/* Xtensa processor ELF architecture-magic number */
 
-	// EM_XTENSA is already definded
+	// EM_XTENSA is already defined
 	public static final int EM_XTENSA_OLD =	0xABC7;
 
 	/* Xtensa relocations defined by the ABIs */
@@ -35,6 +35,7 @@ public class Xtensa_ElfRelocationConstants {
 	public static final int R_XTENSA_OP2 = 10;
 	public static final int R_XTENSA_ASM_EXPAND = 11;
 	public static final int R_XTENSA_ASM_SIMPLIFY = 12;
+	public static final int R_XTENSA_32_PCREL = 14;
 	public static final int R_XTENSA_GNU_VTINHERIT = 15;
 	public static final int R_XTENSA_GNU_VTENTRY = 16;
 	public static final int R_XTENSA_DIFF8 = 17;
@@ -70,4 +71,17 @@ public class Xtensa_ElfRelocationConstants {
 	public static final int R_XTENSA_SLOT12_ALT = 47;
 	public static final int R_XTENSA_SLOT13_ALT = 48;
 	public static final int R_XTENSA_SLOT14_ALT = 49;
+	public static final int R_XTENSA_TLSDESC_FN = 50;
+	public static final int R_XTENSA_TLSDESC_ARG = 51;
+	public static final int R_XTENSA_TLS_DTPOFF = 52;
+	public static final int R_XTENSA_TLS_TPOFF = 53;
+	public static final int R_XTENSA_TLS_FUNC = 54;
+	public static final int R_XTENSA_TLS_ARG = 55;
+	public static final int R_XTENSA_TLS_CALL = 56;
+	public static final int R_XTENSA_PDIFF8 = 57;
+	public static final int R_XTENSA_PDIFF16 = 58;
+	public static final int R_XTENSA_PDIFF32 = 59;
+	public static final int R_XTENSA_NDIFF8 = 60;
+	public static final int R_XTENSA_NDIFF16 = 61;
+	public static final int R_XTENSA_NDIFF32 = 62;
 }
diff --git a/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationHandler.java b/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationHandler.java
index 279ebe4e1..2d0e2e135 100644
--- a/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationHandler.java
+++ b/Ghidra/Processors/Xtensa/src/main/java/ghidra/app/util/bin/format/elf/relocation/Xtensa_ElfRelocationHandler.java
@@ -4,9 +4,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *      http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -15,10 +15,15 @@
  */
 package ghidra.app.util.bin.format.elf.relocation;
 
-import ghidra.app.util.bin.format.elf.*;
+import ghidra.app.util.bin.format.elf.ElfConstants;
+import ghidra.app.util.bin.format.elf.ElfHeader;
+import ghidra.app.util.bin.format.elf.ElfRelocation;
+import ghidra.app.util.bin.format.elf.ElfSymbol;
 import ghidra.program.model.address.Address;
 import ghidra.program.model.listing.Program;
+import ghidra.program.model.mem.Memory;
 import ghidra.program.model.mem.MemoryAccessException;
+import ghidra.program.model.reloc.Relocation.Status;
 import ghidra.program.model.reloc.RelocationResult;
 import ghidra.util.exception.NotFoundException;
 
@@ -26,14 +31,13 @@ public class Xtensa_ElfRelocationHandler extends ElfRelocationHandler {
 
 	@Override
 	public boolean canRelocate(ElfHeader elf) {
-		return elf.e_machine() == ElfConstants.EM_XTENSA ||
-			elf.e_machine() == Xtensa_ElfRelocationConstants.EM_XTENSA_OLD;
+		return elf.e_machine() == ElfConstants.EM_XTENSA
+				|| elf.e_machine() == Xtensa_ElfRelocationConstants.EM_XTENSA_OLD;
 	}
 
 	@Override
-	public RelocationResult relocate(ElfRelocationContext elfRelocationContext,
-			ElfRelocation relocation, Address relocationAddress)
-			throws MemoryAccessException, NotFoundException {
+	public RelocationResult relocate(ElfRelocationContext elfRelocationContext, ElfRelocation relocation,
+			Address relocationAddress) throws MemoryAccessException, NotFoundException {
 		ElfHeader elf = elfRelocationContext.getElfHeader();
 		if (!canRelocate(elf)) {
 			return RelocationResult.FAILURE;
@@ -45,6 +49,7 @@ public class Xtensa_ElfRelocationHandler extends ElfRelocationHandler {
 		}
 
 		Program program = elfRelocationContext.getProgram();
+		Memory memory = program.getMemory();
 
 		ElfSymbol sym = null;
 		String symbolName = null;
@@ -58,152 +63,280 @@ public class Xtensa_ElfRelocationHandler extends ElfRelocationHandler {
 			symbolName = sym.getNameAsString();
 		}
 
-		//int byteLength = 4; // most relocations affect 4-bytes (change if different)
+		int addend = (int) relocation.getAddend();
+		int symbolValue = (int) elfRelocationContext.getSymbolValue(sym);
+		int offset = (int) relocationAddress.getOffset();
+
+		int byteLength = -1;
+		int newValue;
+
+		int diff_mask = 0;
+		boolean neg = false;
 
 		switch (type) {
-			case Xtensa_ElfRelocationConstants.R_XTENSA_32:
-				markAsWarning(program, relocationAddress, "R_XTENSA_32", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_RTLD:
-				markAsWarning(program, relocationAddress, "R_XTENSA_RTLD", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_GLOB_DAT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_GLOB_DAT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_JMP_SLOT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_JMP_SLOT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_RELATIVE:
-				markAsWarning(program, relocationAddress, "R_XTENSA_RELATIVE", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_PLT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_PLT", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_OP0:
-				markAsWarning(program, relocationAddress, "R_XTENSA_OP0", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_OP1:
-				markAsWarning(program, relocationAddress, "R_XTENSA_OP1", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_OP2:
-				markAsWarning(program, relocationAddress, "R_XTENSA_OP2", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_ASM_EXPAND:
-				markAsWarning(program, relocationAddress, "R_XTENSA_ASM_EXPAND", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_ASM_SIMPLIFY:
-				markAsWarning(program, relocationAddress, "R_XTENSA_ASM_SIMPLIFY", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_GNU_VTINHERIT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_GNU_VTINHERIT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_GNU_VTENTRY:
-				markAsWarning(program, relocationAddress, "R_XTENSA_GNU_VTENTRY", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF8:
-				markAsWarning(program, relocationAddress, "R_XTENSA_DIFF8", symbolName, symbolIndex,
-					"TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF16:
-				markAsWarning(program, relocationAddress, "R_XTENSA_DIFF16", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF32:
-				markAsWarning(program, relocationAddress, "R_XTENSA_DIFF32", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT0_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT0_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT1_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT1_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT2_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT2_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT3_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT3_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT4_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT4_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT5_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT5_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT6_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT6_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT7_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT7_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT8_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT8_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT9_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT9_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT10_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT10_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT11_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT11_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT12_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT12_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT13_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT13_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT14_OP:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT14_OP", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT0_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT0_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT1_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT1_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT2_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT2_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT3_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT3_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT4_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT4_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT5_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT5_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT6_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT6_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT7_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT7_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT8_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT8_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT9_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT9_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT10_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT10_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT11_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT11_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT12_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT12_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT13_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT13_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT14_ALT:
-				markAsWarning(program, relocationAddress, "R_XTENSA_SLOT14_ALT", symbolName,
-					symbolIndex, "TODO, needs support ", elfRelocationContext.getLog());
-			default:
-				markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName,
+		case Xtensa_ElfRelocationConstants.R_XTENSA_32:
+			newValue = memory.getInt(relocationAddress);
+			newValue += (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_RTLD:
+			markAsWarning(program, relocationAddress, "R_XTENSA_RTLD", symbolName, symbolIndex, "TODO, needs support ",
+					elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_GLOB_DAT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_GLOB_DAT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_JMP_SLOT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_JMP_SLOT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_RELATIVE:
+			newValue = ((int)elfRelocationContext.getImageBaseWordAdjustmentOffset() + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_PLT:
+			newValue = (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_OP0:
+			markAsWarning(program, relocationAddress, "R_XTENSA_OP0", symbolName, symbolIndex, "TODO, needs support ",
 					elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_OP1:
+			markAsWarning(program, relocationAddress, "R_XTENSA_OP1", symbolName, symbolIndex, "TODO, needs support ",
+					elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_OP2:
+			markAsWarning(program, relocationAddress, "R_XTENSA_OP2", symbolName, symbolIndex, "TODO, needs support ",
+					elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_ASM_EXPAND:
+			markAsWarning(program, relocationAddress, "R_XTENSA_ASM_EXPAND", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_ASM_SIMPLIFY:
+			markAsWarning(program, relocationAddress, "R_XTENSA_ASM_SIMPLIFY", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_32_PCREL:
+			newValue = (symbolValue + addend - offset);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_GNU_VTINHERIT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_GNU_VTINHERIT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_GNU_VTENTRY:
+			markAsWarning(program, relocationAddress, "R_XTENSA_GNU_VTENTRY", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF8:
+			diff_mask = 0x7f;
+			byteLength = 1;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF16:
+			diff_mask = 0x7fff;
+			byteLength = 2;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_DIFF32:
+			diff_mask = 0x7fffffff;
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT0_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT0_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT1_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT1_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT2_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT2_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT3_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT3_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT4_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT4_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT5_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT5_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT6_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT6_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT7_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT7_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT8_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT8_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT9_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT9_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT10_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT10_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT11_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT11_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT12_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT12_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT13_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT13_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT14_OP:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT14_OP", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT0_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT0_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT1_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT1_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT2_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT2_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT3_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT3_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT4_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT4_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT5_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT5_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT6_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT6_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT7_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT7_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT8_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT8_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT9_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT9_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT10_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT10_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT11_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT11_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT12_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT12_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT13_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT13_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_SLOT14_ALT:
+			markAsWarning(program, relocationAddress, "R_XTENSA_SLOT14_ALT", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLSDESC_FN:
+			newValue = (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLSDESC_ARG:
+			newValue = (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLS_DTPOFF:
+			newValue = (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLS_TPOFF:
+			newValue = (symbolValue + addend);
+			memory.setInt(relocationAddress, newValue);
+			byteLength = 4;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLS_FUNC:
+			markAsWarning(program, relocationAddress, "R_XTENSA_TLS_FUNC", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLS_ARG:
+			markAsWarning(program, relocationAddress, "R_XTENSA_TLS_ARG", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_TLS_CALL:
+			markAsWarning(program, relocationAddress, "R_XTENSA_TLS_CALL", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_NDIFF8:
+			neg = true;
+			// fall-through
+		case Xtensa_ElfRelocationConstants.R_XTENSA_PDIFF8:
+			diff_mask = 0xff;
+			byteLength = 1;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_NDIFF16:
+			neg = true;
+			// fall-through
+		case Xtensa_ElfRelocationConstants.R_XTENSA_PDIFF16:
+			diff_mask = 0xffff;
+			byteLength = 2;
+			break;
+		case Xtensa_ElfRelocationConstants.R_XTENSA_NDIFF32:
+			neg = true;
+			// fall-through
+		case Xtensa_ElfRelocationConstants.R_XTENSA_PDIFF32:
+			diff_mask = 0xffffffff;
+			byteLength = 4;
+			break;
+		default:
+			markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, elfRelocationContext.getLog());
+			break;
+		}
+
+		if (diff_mask != 0) {
+			//TODO  not sure if they can all be done here, handle DIFF relocations
+			//SEE   diff_mask, neg, and byteLength
+			markAsWarning(program, relocationAddress, "R_XTENSA_*DIFF*", symbolName, symbolIndex,
+					"TODO, needs support ", elfRelocationContext.getLog());
+			byteLength = -1;
+		}
+
+		if (byteLength < 0) {
+			return RelocationResult.UNSUPPORTED;
 		}
-		return RelocationResult.UNSUPPORTED;
+		return new RelocationResult(Status.APPLIED, byteLength);
 	}
 
 }
-- 
2.43.0

